pandas/io/formats/info.py

```diff
@@ -72,92 +70,114 @@ def _sizeof_fmt(num: Union[int, float], size_qualifier: str) -> str:
     return f"{num:3.1f}{size_qualifier} PB"
 
 
-class BaseInfo(metaclass=ABCMeta):
+class BaseInfo(ABC):
+    """Base class for DataFrameInfo and SeriesInfo.
+
+    Parameters
+    ----------
+    data : FrameOrSeries
+        Either dataframe or series.
+    memory_usage : bool or str, optional
+        If "deep", introspect the data deeply by interrogating object dtypes
+        for system-level memory consumption, and include it in the returned
+        values.
+    """
+
     def __init__(
         self,
         data: FrameOrSeries,
```

Comment:
:thinking: not sure if `FrameOrSeries` is correct here, as the comment near it says
```
# FrameOrSeriesUnion  means either a DataFrame or a Series. E.g.
# `def func(a: FrameOrSeriesUnion) -> FrameOrSeriesUnion: ...` means that if a Series
# is passed in, either a Series or DataFrame is returned, and if a DataFrame is passed
# in, either a DataFrame or a Series is returned.
```
, which `__init__` doesn't return anything. However, `FrameOrSeriesUnion` also talks about either a Series or a DataFrame being returned, so I'm not sure how to type this if nothing is returned... @topper-123 any suggestions?

Reply:
Looking at the documentation, I guess ``FrameOrSeries`` means the following.
Consider ``x: FrameOrSeries``.
Function ``f(x) -> FrameOrSeries``.
Let us say, that now ``x`` is a dataframe, then ``f(x)`` must return ``DataFrame``, but not ``Series``.
Meanwhile ``FrameOrSeriesUnion`` allows that ``f(x)`` returns different type.

Reply:
Yes, that's my understanding too, although technically `__init__` doesn't return anything, so...just `Union[Series, DataFrame]`?

Reply:
``FrameOrSeriesUnion = Union["DataFrame", "Series"]``, I can use it.

It we use ``FrameOrSeries``, then it does not mean that we must return.
It just means that **if** we return, we **must** return the same type.

It is not like I object against changes.
I only mean that these are not required.
mypy would raise issues if any.

Reply:
I tried ``FrameOrSeriesUnion`` and it leads to mypy complains.

```
pandas/io/formats/info.py:140: error: Incompatible return value type (got "Union[Any, ExtensionDtype]", expected "Series")  [return-value]
pandas/io/formats/info.py:201: error: Incompatible return value type (got "Union[Any, ExtensionDtype]", expected "Series")  [return-value]
```

So, I leave it as it is for now.

Reply:
I think you're right, e.g. `pandas/io/pytables.py` contains:
```python
def to_hdf(
    path_or_buf,
    key: str,
    value: FrameOrSeries,
    mode: str = "a",
    complevel: Optional[int] = None,
    complib: Optional[str] = None,
    append: bool = False,
    format: Optional[str] = None,
    index: bool = True,
    min_itemsize: Optional[Union[int, Dict[str, int]]] = None,
    nan_rep=None,
    dropna: Optional[bool] = None,
    data_columns: Optional[Union[bool, List[str]]] = None,
    errors: str = "strict",
    encoding: str = "UTF-8",
):
```

Let's keep as is then, thanks for explaining!